samples/outreach/blogs/serving_blogpost/data_utils.py

```diff
@@ -0,0 +1,204 @@
+import numpy as np
+import tensorflow as tf
+from tensorflow.python.keras import backend as K
+import functools
+import matplotlib.pyplot as plt
+import urllib.request
+from io import BytesIO
+from zipfile import ZipFile
+import urllib.request
+import os
+
+def get_all_data():
+  directory = './data'
+  if not os.path.exists(directory):
+    os.makedirs(directory)
+
+  base_url = 'https://www.bgc-jena.mpg.de/wetter/mpi_roof_20{:02d}{}.zip'
+
+  merge_2006_2012 = []
+  for i in range(6, 12):
+    merge_2006_2012.append(base_url.format(i, 'a'))
+    merge_2006_2012.append(base_url.format(i, 'b'))
+
+  get_set_of_data(merge_2006_2012, os.path.join(directory, 'merge_2006_2012.csv'))
+
+  merge_2012_2018 = []
+  for i in range(12, 18):
+    merge_2012_2018.append(base_url.format(i, 'a'))
+    merge_2012_2018.append(base_url.format(i, 'b'))
+  # Special case for present data
+  merge_2012_2018.append('https://www.bgc-jena.mpg.de/wetter/mpi_roof.zip')
+
+  get_set_of_data(merge_2012_2018, os.path.join(directory, 'merge_2012_2018.csv'))
+
+
+def get_set_of_data(urls, output_filename):
+  output_file = open(output_filename, 'wb')
+  for i, url in enumerate(urls):
+    print('Downloading: {}'.format(url))
+    get_single_zip_data(url, output_file, i == 0)
+
+  output_file.close()
+
+
+def get_single_zip_data(url, output_file, first=False):
+  url = urllib.request.urlopen(url)
+  with ZipFile(BytesIO(url.read())) as my_zip_file:
+    contained_file = my_zip_file.namelist()[0]
+    for i, line in enumerate(my_zip_file.open(contained_file).readlines()):
+      if i == 0 and not first or '-9999' in str(line):
+        # Skip in the case of a header (and not the first file) or invalid data
+        continue
+      output_file.write(line)
+
+class Dataset():
+  def __init__(self,
+               data_path='./data/merge_2006_2012.csv',
+               batch_size=128,
+               data_points_day=144,
+               num_days=10,
+               steps=6,
+               mode='train'):
+    self.num_days = num_days
+    self.data_points_day = data_points_day
+    self.data_path = data_path
+    # Look back 10 days
+    self.lookback = data_points_day * num_days
+    # observations once every hour
+    self.steps = steps
+    # We will target one day in the future (24 hours)
+    self.delay = data_points_day
+    self.batch_size = batch_size
+    if mode == 'inference':
+      print('Setting batch size to 1 in inference mode')
+      self.batch_size = 1
+
+  def _preprocess_dataset_element(self, idx, data, lookback, steps, delay):
+    """fn to return the actual data given a specific index"""
+    # We map this function onto individual elements of the dataset
+    sample = data[idx - lookback: idx: steps]
```

Comment:
`steps` in this case isn't a number of steps, it's the sampling frequency, so that we sample at one data point per hour (with each time step every 10 minutes). The default value of `step` is 6, meaning we grab at every 6th element. Does that sound correct?

Reply:
Ah, I should have read more closely, you explain this in the comments. 
This looks correct as is.